### PR TITLE
[NO SQUASH] Non-SDL: Add opengl3 support

### PR DESCRIPTION
**Pull request by SmallJoker**
`video_driver = opengl3` support for the non-SDL Linux and Win32 Irrlicht device.

Benefits:

* Less code duplication (of stub functions)
* Alternative access to OpenGL 3 without the SDL2 dependency
* Minimal amount of additional code.
* ????

The change `GL_VENDOR` -> `GL_RENDERER` was done because latter does provide more helpful information about the video card that is being used. See `debug_log_level = info` + `debug.txt` for comparison.

* Before: `glxinfo | grep "OpenGL vendor string"` (vendor name only)
* Now: `glxinfo | grep "OpenGL renderer string"` (more detailed graphics card information)

## To do

This PR is Ready for Review.

## How to test

1. Compile with `USE_SDL2=0` and `ENABLE_OPENGL3=1`
2. See window title
3. See debug.txt
4. Shaders must work
5. Dynamic shadows should work properly
6. ??

### CODE REVIEW

File: irr/src/CEGLManager.cpp [hunk 1]
```diff
@@ -156,6 +198,8 @@ EGLConfig CEGLManager::chooseConfig(EConfigStyle confStyle)
 	case EDT_WEBGL1:
 		eglOpenGLBIT = EGL_OPENGL_ES2_BIT;
 		break;
+	case EDT_OPENGL:
```
Comment (sfan5) [inline irr/src/CEGLManager.cpp:201]: missing gl3 here?

File: irr/src/CEGLManager.cpp [hunk 2]
```diff
@@ -156,6 +198,8 @@ EGLConfig CEGLManager::chooseConfig(EConfigStyle confStyle)
...
+		eglOpenGLBIT = EGL_OPENGL_BIT;
```
Comment (sfan5) [inline irr/src/CEGLManager.cpp:202]: (I forgot this)
```suggestion
		eglOpenGLBIT = EGL_OPENGL_BIT;
		break;
```

**Comment by SmallJoker** [comment]
The rebase was trivial. I am continuing this until SDL is mandatory on all platforms.

File: irr/README.md [hunk 3]
```diff
@@ -36,6 +36,28 @@ We aim to support these platforms:
...
+| WebGL 1                   | Untested | Untested (EGL) | Not implemented  | Not implemented |
+| Null (no graphics output) | Works    | Works          | Works            | Works           |
+
+Notes:
+
+* [1] `CIrrDeviceSDL`: supports Android, Linux, macOS, Windows
+* [2] `CIrrDeviceLinux`: supports Linux
+* [3] `CIrrDeviceOSX`: supports macOS
+* [4] `CIrrDeviceWin32`: supports Windows
+* [5] Newer versions are compatible. New features and functions are not used.
```
Comment (sfan5) [inline irr/README.md:58]: if we want to be exact that isn't any different for the other drivers

File: src/client/shadows/dynamicshadowsrender.cpp [hunk 4]
```diff
@@ -694,9 +694,13 @@ std::string ShadowRenderer::readShaderFile(const std::string &path)
...
 {
 	// disable if unsupported
-	if (g_settings->getBool("enable_dynamic_shadows") &&
-		device->getVideoDriver()->getDriverType() != video::EDT_OPENGL) {
-		g_settings->setBool("enable_dynamic_shadows", false);
+	if (g_settings->getBool("enable_dynamic_shadows")) {
+		// See also checks in builtin/mainmenu/settings/dlg_settings.lua
+		const video::E_DRIVER_TYPE type = device->getVideoDriver()->getDriverType();
+		if (!(type == video::EDT_OPENGL || type == video::EDT_OPENGL3)) {
+			warningstream << "Shadows: disabled dynamic shadows due to missing dependencies" << std::endl;
```
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.cpp:701]:
```suggestion
			warningstream << "Shadows: disabled dynamic shadows due to being unsupported" << std::endl;
```

File: irr/src/CMakeLists.txt [hunk 5]
```diff
@@ -152,8 +152,10 @@ else()
 	endif()
 endif()
 
-if(ENABLE_OPENGL)
-	add_definitions(-D_IRR_COMPILE_WITH_OPENGL_)
+if(ENABLE_OPENGL OR (ENABLE_OPENGL3 AND NOT USE_SDL2))
+	if(ENABLE_OPENGL)
+		add_definitions(-D_IRR_COMPILE_WITH_OPENGL_)
+	endif()
 	set(OPENGL_DIRECT_LINK TRUE) # driver relies on this
```
Comment (sfan5) [inline irr/src/CMakeLists.txt:159]: should be inside `if(ENABLE_OPENGL)`

File: irr/src/OpenGL/Driver.cpp [hunk 6]
```diff
@@ -226,7 +226,7 @@ void COpenGL3DriverBase::initVersion()
 	printVersion();
 
 	// print renderer information
-	VendorName = GL.GetString(GL_VENDOR);
+	VendorName = GL.GetString(GL_RENDERER);
 	os::Printer::log("Vendor", VendorName.c_str(), ELL_INFORMATION);
```
Comment (sfan5) [inline irr/src/OpenGL/Driver.cpp:230]:
```suggestion
	os::Printer::log("Renderer", VendorName.c_str(), ELL_INFORMATION);
```

File: src/client/shadows/dynamicshadowsrender.cpp [hunk 7]
```diff
@@ -694,9 +694,13 @@ std::string ShadowRenderer::readShaderFile(const std::string &path)
 ShadowRenderer *createShadowRenderer(IrrlichtDevice *device, Client *client)
 {
 	// disable if unsupported
-	if (g_settings->getBool("enable_dynamic_shadows") &&
-		device->getVideoDriver()->getDriverType() != video::EDT_OPENGL) {
-		g_settings->setBool("enable_dynamic_shadows", false);
+	if (g_settings->getBool("enable_dynamic_shadows")) {
+		// See also checks in builtin/mainmenu/settings/dlg_settings.lua
+		const video::E_DRIVER_TYPE type = device->getVideoDriver()->getDriverType();
+		if (!(type == video::EDT_OPENGL || type == video::EDT_OPENGL3)) {
```
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.cpp:700]:
```suggestion
		const auto type = device->getVideoDriver()->getDriverType();
		if (type != video::EDT_OPENGL && type != video::EDT_OPENGL3) {
```

File: irr/src/CMakeLists.txt (hunk 5 above)
Comment (SmallJoker) [reply irr/src/CMakeLists.txt:159]: Hmm. In my testing, this made no difference. Why exactly is it needed, and for which code?

Comment (sfan5) [reply irr/src/CMakeLists.txt:159]: the legacy GL driver needs it, but not the GL3 driver.

**Review by sfan5 — APPROVED** [review]
doesn't break anything. shadows work on gl3.

**Pull request merged**